subprojects/execution/src/main/java/org/gradle/internal/execution/fingerprint/FileCollectionSnapshotter.java

```diff
@@ -23,8 +23,21 @@
  * Service for snapshotting {@link FileCollection}s.
  */
 public interface FileCollectionSnapshotter {
+    interface Result {
+        FileSystemSnapshot getSnapshot();
+
+        /**
+         * Whether or not the snapshotted file collection consists only of file trees.
+         *
+         * If the file collection does not contain any file trees, then this will return {@code false}.
+         */
+        boolean isFileTreeOnly();
+    }
```

Comment:
The main problem here is that I can't do the visiting in `DefaultInputFingerprinter`, since `execution` doesn't know about `FileCollectionInternal.visitStructure`. So I'd need to put another service in between.

Reply:
Alright, let's leave it like this then.